### ISSUE TITLE
yggtorrent engine: Fail to initialize / does not work

### ISSUE THREAD

**Issue by return42** [searx/searx#2612]
**Version of Searx, commit number if you are using on master branch and stipulate if you forked Searx**

branch master

**How did you install Searx?**

    make run

Always get a _yggtorrent engine: Fail to initialize_ / e.g. [here in every build](https://github.com/searx/searx/pull/2611/checks#step:5:297))::

    ERROR:searx.engines:yggtorrent engine: Fail to initialize
    ....
    File "/home/runner/work/searx/searx/searx/engines/yggtorrent.py", line 42, in init
        resp = http_get(url, allow_redirects=False)
    ...
    raise SearxEngineCaptchaException(message='Cloudflare CAPTCHA', suspended_time=3600 * 24 * 15)

The cloudflare captcha comes from the init function of yggtorrent:

https://github.com/searx/searx/blob/e2fb50089219207f7e11bcfc049b48f2f4b9a6c8/searx/engines/yggtorrent.py#L39-L46

@unixfox do you have a clue how we can bypass  the captcha (at least in the init process)?

BTW: CAPTCHA problem is not limited to init proc / all `!ygg` queries result in _yggtorrent (CAPTCHA required)_ or even _Sorry!_
 after first CAPTCHA message.

**Comment by unixfox** [comment]
Well yggtorrent sometimes enable their cloudflare captcha but other times it's not activated. It depends from day to day.
What I could do it's to check if I get a captcha and then stop the init function. I didn't write it this way at first because there was no captcha at that moment.

By the way that's something that could happen one day on any existing engine with an init function. When thinking about it this way, I don't know if it's better to take into account the captcha inside the engine code or generally in the core of Searx.

**Comment by return42** [comment]
> What I could do it's to check if I get a captcha and then stop the init function. 

Sounds like a good solution .. do you like to prepare a PR (otherwise I can do)

> I don't know if it's better to take into account the captcha inside the engine code or generally in the core of Searx.

IMO the `SearxEngineCaptchaException` needs to be handled where it is raised .. here it is raised the first time in the init function. So I would prefer _"explicit"_ over _"general"_.

**Comment by dalf** [comment]
The current engine initialization:
https://github.com/searx/searx/blob/e2fb50089219207f7e11bcfc049b48f2f4b9a6c8/searx/engines/__init__.py#L291-L297

If instead of `except Exception`, the [online processor](https://github.com/searx/searx/blob/e2fb50089219207f7e11bcfc049b48f2f4b9a6c8/searx/search/processors/online.py#L158-L217) handles the exception, it would suspend the engine for a time. Related to [abstract.py](https://github.com/dalf/searx/blob/66d826a8f6ccf7ae700f964a8d5518f5dfbc04f6/searx/search/processors/abstract.py) from https://github.com/searx/searx/pull/2561

Said differently it raises the question to move the engine initialization to the processor module: in addition to the `def search(self, query, params, result_container, start_time, timeout_limit)`, the abstract processor could have a `def initialize(self, settings)` method.

It makes even more sense looking at [_set_https_support_for_engine](https://github.com/searx/searx/blob/e2fb50089219207f7e11bcfc049b48f2f4b9a6c8/searx/engines/__init__.py#L309-L335).

To sum up:
* the processor catches the exception and stop any request for a time (by default one day, can be changed in the SearxEngineCaptchaException).
* the exception wouldn't show up in the log except in debug mode.

**Comment by return42** [comment]
> To sum up:  ... the exception wouldn't show up in the log except in debug mode.

not in the log, but on stderr ...

```
./local/py3/bin/python -c "import searx.search; searx.search.initialize()"
ERROR:searx.engines:yggtorrent engine: Fail to initialize
Traceback (most recent call last):
  File "searx/searx/engines/__init__.py", line 293, in engine_init
    init_fn(get_engine_from_settings(engine_name))
...
searx.exceptions.SearxEngineCaptchaException: Cloudflare CAPTCHA, suspended_time=1296000
```

**Comment by dalf** [comment]
> not in the log, but on stderr ...

In this case, in this case, in this code:

https://github.com/searx/searx/blob/0165e14a7f0400ae10bfb217e4e4bf3992db6bf1/searx/__init__.py#L50-L53

`logging.basicConfig` should have additional parameter `filename='/dev/stdout'` or better, use the `stream` argument.

**Comment by return42** [comment]
> By the way that's something that could happen one day on any existing engine with an init function. When thinking about it this way, I don't know if it's better to take into account the captcha inside the engine code or generally in the core of Searx.

I think you are right .. it could be a normal use case ... and it should be handled in the core of searx .. see my PR #2635.
